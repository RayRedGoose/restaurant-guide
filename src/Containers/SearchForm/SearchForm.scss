@import "../../assets/css/variables.scss";
@import "../../assets/css/index.scss";

.search-form {
  display: flex;
  position: relative;
  width: 30%;
  input[type="text"] {
    border-bottom: 2px solid $box-color;
    transition: width 2s;
    padding: 0.3em 1em;
    &.opened {
      width: 100%;
    }
    &.closed {
      width: 25%;
    }
  }
  .none {
    display: none;
  }
  .remove-btn {
    cursor: pointer;
    position: absolute;
    top: 0.5rem;
    width: 0.5rem;
    opacity: 0.5;
    &:hover {
      opacity: 1;
    }
  }
  .search-btn {
    display: flex;
    align-items: center;
    justify-content: center;
    background-color: $box-color;
    border-radius: 50%;
    color: darken($box-color, 50%);
    margin-left: -1rem;
    padding: 0.5em;
    svg {
      height: 0.85rem;
    }
    &:hover {
      background-color: darken($box-color, 50%);
      color: $background-color;
    }
  }
}
