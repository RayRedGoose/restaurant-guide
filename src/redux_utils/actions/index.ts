import {
  IRestaurantObject,
  IAddRestaurantAction
} from 'assets/ts/interfaces';

export const addRestaurants = (restaurants: IRestaurantObject[]): IAddRestaurantAction => ({
  type: 'ADD_RESTAURANTS',
  restaurants: restaurants
});
