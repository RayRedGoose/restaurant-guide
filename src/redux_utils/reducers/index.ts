import { combineReducers } from 'redux';
import restaurants from './restaurants/restaurants';

const rootReducer = combineReducers({
  restaurants
});

export default rootReducer;
