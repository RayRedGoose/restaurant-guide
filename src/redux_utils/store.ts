import { Store, createStore } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import rootReducer from "./reducers";
import { IAppStore } from "assets/ts/interfaces";

const store: Store<IAppStore> = createStore(rootReducer, composeWithDevTools());

export default store;
