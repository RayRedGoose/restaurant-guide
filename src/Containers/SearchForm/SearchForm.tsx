import "./SearchForm.scss";
import React, { useState, useEffect, FormEvent, ChangeEvent } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSearch, faTimes } from "@fortawesome/free-solid-svg-icons";
import { useDispatch, useSelector } from "react-redux";
import { addSearchFilter, removeSearchFilter } from "redux_utils/actions";
import { IAppStore } from "assets/ts/interfaces";

const SearchForm: React.FC = () => {
  const dispatch = useDispatch();
  const [query, setQuery] = useState<string>("");
  const [isOpened, setIsOpened] = useState<boolean>(false);
  const searchFilter = useSelector((state: IAppStore) => state.searchFilter);

  useEffect(() => {
    setQuery(searchFilter);
  }, [searchFilter]);

  const submitSearch = (e: FormEvent<HTMLElement>): void => {
    e.preventDefault();
    if (query !== "") dispatch(addSearchFilter(query));
  };

  const handleChanges = (e: ChangeEvent<HTMLInputElement>): void => {
    setQuery(e.target.value);
    if (e.target.value === "") dispatch(removeSearchFilter());
  };

  const cleanInput = (): void => {
    setQuery("");
    dispatch(removeSearchFilter());
  };

  return (
    <form className="search-form" onSubmit={submitSearch}>
      {query !== "" && (
        <FontAwesomeIcon
          className="remove-btn"
          icon={faTimes}
          onClick={cleanInput}
          onKeyPress={cleanInput}
          tabIndex={0}
        />
      )}
      <input
        type="text"
        placeholder="Search"
        value={query}
        onChange={handleChanges}
        onFocus={() => setIsOpened(true)}
        onBlur={() => setIsOpened(false)}
        className={!isOpened && query === "" ? "closed" : "opened"}
      />
      <button className="search-btn" onClick={submitSearch}>
        <FontAwesomeIcon icon={faSearch} />
      </button>
    </form>
  );
};

export default SearchForm;
