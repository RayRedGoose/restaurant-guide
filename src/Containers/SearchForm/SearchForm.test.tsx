/**
 * @jest-environment jsdom
 */

import React from 'react';
import { shallow, mount } from 'enzyme';
import SearchForm from './SearchForm';
import { addSearchFilter, removeSearchFilter } from 'redux_utils/actions';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useDispatch: () => mockDispatch
}));

jest.mock('redux_utils/actions');

describe("SearchForm", () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = mount(<SearchForm />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update input state when change event occur', () => {
    const input = wrapper.find('input');
    expect(input.getDOMNode().value).toEqual('');

    input.simulate('change', {
      target: { value: 'hi!' }
    });
    expect(input.getDOMNode().value).toEqual('hi!');
  });

  it('should removeSearchFilter when change event fired with empty value', () => {
    const input = wrapper.find('input');
    input.simulate('change', {
      target: { value: '' }
    });
    expect(removeSearchFilter).toHaveBeenCalled();
  });

  it('should addSearchFilter when form is submitted', () => {
    wrapper.find('input').simulate('change', {
      target: { value: 'query' }
    });
    wrapper.find('form').simulate('submit');
    expect(addSearchFilter).toHaveBeenCalledWith('query');
  });

  it('should addSearchFilter when submit button is clicked', () => {
    wrapper.find('input').simulate('change', {
      target: { value: 'query' }
    });
    wrapper.find('button').simulate('click');
    expect(addSearchFilter).toHaveBeenCalledWith('query');
  });
});
