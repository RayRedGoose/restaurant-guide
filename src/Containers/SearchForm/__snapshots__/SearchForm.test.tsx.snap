// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should match the snapshot 1`] = `
<SearchForm>
  <form
    className="search-form"
    onSubmit={[Function]}
  >
    <input
      className="closed"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
    <button
      className="search-btn"
      onClick={[Function]}
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [],
              "f002",
              "M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z",
            ],
            "iconName": "search",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-search fa-w-16 "
          data-icon="search"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </FontAwesomeIcon>
    </button>
  </form>
</SearchForm>
`;

exports[`SearchForm should match the snapshot if input is not empty 1`] = `
<SearchForm>
  <form
    className="search-form"
    onSubmit={[Function]}
  >
    <FontAwesomeIcon
      border={false}
      className="remove-btn"
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            352,
            512,
            Array [],
            "f00d",
            "M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z",
          ],
          "iconName": "times",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      onKeyPress={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      tabIndex={0}
      title=""
      transform={null}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-times fa-w-11 remove-btn"
        data-icon="times"
        data-prefix="fas"
        focusable="false"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="img"
        style={Object {}}
        tabIndex={0}
        viewBox="0 0 352 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </FontAwesomeIcon>
    <input
      className="opened"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Search"
      type="text"
      value="query"
    />
    <button
      className="search-btn"
      onClick={[Function]}
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [],
              "f002",
              "M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z",
            ],
            "iconName": "search",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-search fa-w-16 "
          data-icon="search"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </FontAwesomeIcon>
    </button>
  </form>
</SearchForm>
`;
